Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Dashboard/Driver.hs

```diff
@@ -925,19 +928,52 @@ addVehicleForFleet merchantShortId opCity reqDriverPhoneNo mbMobileCountryCode f
   vehicle <- RCQuery.findLastVehicleRCWrapper req.registrationNo
-  whenJust vehicle $ \veh -> when (isJust veh.fleetOwnerId && veh.fleetOwnerId /= Just fleetOwnerId) $ throwError VehicleBelongsToAnotherFleet
+  whenJust vehicle $ \veh -> do
+    when (isJust veh.fleetOwnerId && veh.fleetOwnerId /= Just fleetOwnerId) $ throwError VehicleBelongsToAnotherFleet
+    activeAssociationsOfRC <- DRCAE.findAllActiveAssociationByRCId veh.id
+    let rcAssociatedDriverIds = map (.driverId) activeAssociationsOfRC
+    activeFleetDrivers <- FDV.findAllDriverByFleetOwnerId' fleetOwnerId
+    let fleetDriverIds = map (.driverId) activeFleetDrivers
+    let isActiveWithNonFleetDriver = any (`notElem` fleetDriverIds) rcAssociatedDriverIds
+    when isActiveWithNonFleetDriver $ throwError (InvalidRequest "Vehicle is associated with a driver who is not part of this fleet, First Unlink the vehicle from that driver and then try again")
   isFleetDriver <- FDV.findByDriverIdAndFleetOwnerId driver.id fleetOwnerId
   case isFleetDriver of
     Nothing -> throwError (InvalidRequest "Driver is not part of this fleet, add this driver to the fleet before adding a vehicle with them")
     Just fleetDriver -> do
       unless fleetDriver.isActive $ throwError (InvalidRequest "Driver is not active with this fleet, add this driver to the fleet before adding a vehicle with them")
   Redis.set (DomainRC.makeFleetOwnerKey req.registrationNo) fleetOwnerId -- setting this value here , so while creation of creation of vehicle we can add fleet owner id
-  void $ runVerifyRCFlow driver.id merchant merchantOpCityId req
+  void $ runVerifyRCFlow driver.id merchant merchantOpCityId opCity req True
+  logTagInfo "dashboard -> addVehicle : " (show driver.id)
+  pure Success
+
+---------------------------------------------------------------------
+
+addVehicleForFleetWithoutDriver :: ShortId DM.Merchant -> Context.City -> Text -> Common.AddVehicleReq -> Flow APISuccess
+addVehicleForFleetWithoutDriver merchantShortId opCity fleetOwnerId req = do
+  runRequestValidation Common.validateAddVehicleReq req
+  merchant <- findMerchantByShortId merchantShortId
+  merchantOpCityId <- CQMOC.getMerchantOpCityId Nothing merchant (Just opCity)
+  let personId = Id fleetOwnerId :: Id DP.Person
+  driver <-
+    QPerson.findById personId
+      >>= fromMaybeM (PersonDoesNotExist personId.getId)
+  let merchantId = driver.merchantId
+  unless (merchant.id == merchantId && merchantOpCityId == driver.merchantOperatingCityId) $ throwError (PersonDoesNotExist driver.id.getId)
+  vehicle <- RCQuery.findLastVehicleRCWrapper req.registrationNo
+  whenJust vehicle $ \veh -> do
+    when (isJust veh.fleetOwnerId && veh.fleetOwnerId /= Just fleetOwnerId) $ throwError VehicleBelongsToAnotherFleet
+    activeAssociationsOfRC <- DRCAE.findAllActiveAssociationByRCId veh.id
+    let rcAssociatedDriverIds = map (.driverId) activeAssociationsOfRC
+    activeFleetDrivers <- FDV.findAllDriverByFleetOwnerId' fleetOwnerId
```

Comment:
Instead of fetching all drivers for that fleet, isn't it better to check for the RC linked drivers only?

Reply:
yes, but i thought for this we have to do one db call , and for RC linked drivers we can have multiple driver call depending upon the number of driver calls. or is there a way i can do it in one db call ??

Reply:
Its not about one db call, its about how much heavy that call is. With time this table will keep growing, its better to avoid one big single db call.